cc: convert compiler wrappers to POSIX sh

cosmicexplorer opened the pull request:
- Fixes #8703.
- Closes #25380.
- Supersedes #25557.

This PR removes the use of several bash-specific features, in particular the use of arrays and the more featureful `[[ ... ]]` comparison operator. In short:
1. Array operations are replaced with `append()`, `prepend()`, `concatenate()`, etc methods which use the alarm bell ASCII character `'\a'` (`'\007'` in octal) to delimit entries in the array. The alarm bell character is then used as the `IFS` special variable to split up and iterate over these pseudo-arrays.
2. Several verbose `if` comparisons which have gotten crusty over the years have been converted into should-be-easier-to-read `case` statements.
3. `eval` is used in a couple more places instead of bash's indirect variable dereference operator `${!...}`.
4. `$SPACK_TARGET_ARGS` is set to `'\n'` in `cc.py` and `build_environment.py`, since posix `sh` does not have the ability to distinguish whether a variable is unset or set to the empty string.
5. `printf` is used over `echo` in all cases, since `echo` has unpredictable behavior across shells (see "Printing the value of a variable" in https://www.etalabs.net/sh_tricks.html).
    - Relatedly, since `"\n"` does not get encoded into a newline in posix `sh` unless as part of an `echo` or `printf`, there is an annoyingly verbose check to see whether `$SPACK_TARGET_ARGS` is set to a newline. 
6. `set -u` is set at the top of the script, and variables are all explicitly set to a default value if not provided.

### Performance Analysis: 14.5% slowdown

The current changes appear to induce a **14.5% slowdown** in the time to configure `gettext@0.21` compared to `develop`. As per @haampie's strategy in https://github.com/spack/spack/pull/25380#issuecomment-901155651:
```bash
# Tested on ubuntu 21.10, kernel 5.13.0-16, AMD Ryzen 9 5900HX.
# On develop (using /bin/bash -> gnu bash version 5.1.8):
Time =  71.29999876022339 # first run
Time =  70.31215143203735 # second run
# On this branch (using /bin/sh -> dash version 0.5.11):
Time =  81.6425609588623 # first run
Time =  81.04689812660217 # second run
# On this branch (using /bin/bash -> gnu bash version 5.1.8):
Time = 104.45283937454224
```
I *believe* this can be explained mostly by the use of string operations and `IFS` splitting noted above instead of native bash arrays. This suspicion is circumstantiated by noting the final timing number, where I swapped out `#!/bin/sh` for `#!/bin/bash` to see whether using `dash` over `bash` had any negative performance effect. It seems like the opposite occurred, where `bash` was actually significantly slower than `dash` to execute the same script, which might make sense given that `bash` users are more likely to use its native arrays over string operations, and that that's likely where the gnu `bash` project has focused its optimization efforts.

As a final note, it might be useful to consider performance regression testing in the above vein in the future, especially since we can then perform such testing over e.g. macOS and (in the future) windows, instead of the developer's local machine which may be running other interactive applications. For the case of improving `./configure` performance though, we may want to consider looking to this discussion I created recently proposing to allow reusing stage directories across subsequent installs of the same package: #25851.

haampie committed:
cc: make some easy initial posix conversions based on `shellcheck`

Fixes the following shellcheck issues:

* https://github.com/koalaman/shellcheck/wiki/SC3011
* https://github.com/koalaman/shellcheck/wiki/SC3024
* https://github.com/koalaman/shellcheck/wiki/SC3030
* https://github.com/koalaman/shellcheck/wiki/SC3045
* https://github.com/koalaman/shellcheck/wiki/SC3053
* https://github.com/koalaman/shellcheck/wiki/SC3054

The "only" things left to fix are arrays, read, heredocs, and
other some trivial things we can work around

cosmicexplorer committed:
convert the first half to posix sh

cosmicexplorer committed:
convert the second half of the script to posix sh

cosmicexplorer committed:
quote every single assignment (heavy sigh)

cosmicexplorer committed:
make it explicit when a variable is intentionally unset by spack

cosmicexplorer committed:
add valid bool check and re-add a comment line

cosmicexplorer committed:
remove some redundant string manipulation for path comparison

cosmicexplorer committed:
fix file comparison test on macos

cosmicexplorer committed:
style fixes

cosmicexplorer committed:
make it pass shellcheck, and move function definitions to the top

cosmicexplorer committed:
declare every variable as readonly after the final modification

cosmicexplorer committed:
increase the license lines length limit to allow shellcheck ignores

haampie commented:
Thanks for all this, that was quite the shell scripting exercise :smile:.

I'm a bit worried about the overhead :( I've verified this on Ubuntu 20.04 where gettext configure goes from 66.1 seconds without any compiler wrapper to 89.4 seconds, so the overhead of the compiler wrapper is over 35% after this change, which is very significant.

Are generated compiler wrappers the way to go then? It would help us get rid of arrays, loops and string processing

cosmicexplorer committed:
convert SPACK_TARGET_ARGS to use a newline

cosmicexplorer commented:
@haampie I was under the impression that the appropriate performance number for this change would be the %slowdown vs `develop`, which in my testing is only 14.5%. I agree it is not ideal that this produces an additional slowdown vs executing the compiler directly, but I was thinking that that would be a separate problem than conformance to posix. I would love to volunteer to look into ways to improve the performance of this script vs the raw compiler wrapper in a separate change, however.

I've also noted in the OP that #25851 describes an alternative method for speeding up `./configure` phases by retaining the stage directory for a package across installations -- that could address this concern.

haampie commented:
> I was under the impression that the appropriate performance number for this change would be the %slowdown vs develop, which in my testing is only 14.5%

Can confirm, I'm getting 18% from the best of two runs (75.44s -> 89.38s). Even that is quite significant though; anything above 10% is a regression in my subjective views ^^. But we should not neglect the absolute overhead, cause new users are looking at the 35% number, not at the difference with regard to a point in time where they didn't use Spack yet. Maybe in HPC people are used to slow vendor compiler wrappers, but it wouldn't hurt to get something that's actually low overhead :p

That said, `gettext` is just one example I took because of its ~terrible~ extensive configure script. We can maybe collect some other C/Fortran (skipping C++ on purpose, since C++ is slow to compile anyways) packages like `openblas` / `fftw` and do some more benchmarking to get a better picture.

> I've also noted in the OP that #25851 describes an alternative method for speeding up ./configure phases by retaining the stage directory for a package across installations -- that could address this concern.

That's great! But it speeds up the second time only, right? For provisioning a center, I think it's quite common to build from scratch, just to be sure that cache doesn't somehow influence the results in a hard to debug way.

cosmicexplorer commented:
Ok, I figured out the very annoying issue that was happening on macos clingo bootstrapping that was distracting me earlier just now, so I'll look into that performance concern now. Certainly we can consider generating a wrapper for processing things like environment variables which aren't likely to change across invocations of the compiler within a `./configure` script.

cosmicexplorer commented:
If it takes nontrivial effort to solve the performance issue I will demand to have a discussion about performance regression testing before merging this so we don't backslide.

cosmicexplorer committed:
convert '' -> '\n' in another test

cosmicexplorer committed:
convert /usr/bin/[ to use case

cosmicexplorer commented:
I managed to get a tiny speedup by converting invocations of the `[` utility (which calls out to `/usr/bin/[` in `dash`) into mildly-verbose `case` statements. Result:
```bash
Time =  78.90805339813232
```

haampie commented:
FWIW, looking at the `Build` time from spack's logs (which includes more than just compilation):

|                                     | no compiler wrapper | develop        | #26048         |
|-------------------------------------|---------------------|----------------|----------------|
| `openblas`                          | 53.94s, 54.10s      | 57.45s, 57.65s | 75.60s, 75.59s |
| `fftw` | 61.61s, 61.16s      | 63.80s, 64.21s | 70.95s, 71.98s |

So relative to no compiler wrapper it's +6% and +4% for `develop` versus +40% and +16% for this PR. Openblas does a lot of compiler invocations for small files. Also note that openblas / fftw may not be the best examples out there, since they have no link dependencies, so few -Wl,-rpath, -I, -L flags injected by spack, but it does make it easier to run a quick benchmark.

```yaml
spack:
  config:
    ccache: false
    install_tree:
      root: ./here
  specs:
  - openblas
  - fftw ~mpi ~openmp precision=float
  view: false
  packages:
    perl:
      externals:
      - spec: perl@5.30.0~cpanm+shared+threads
        prefix: /usr
```

tgamblin commented:
> I was thinking that that would be a separate problem than conformance to posix. I would love to volunteer to look into ways to improve the performance of this script vs the raw compiler wrapper in a separate change, however.

We likely won't merge this if it slows down builds this much.  The hope was that it would speed things up.  We should figure out what the slowdown is.

cosmicexplorer commented:
Great! Thanks for clarifying.

cosmicexplorer commented:
For fun, I tried writing this in `awk` earlier this weekend. It was fun to learn `awk`, and the result is competitive with the *initial* version of this PR, clocking in at just over 80 seconds (about a second faster than the initial attempt). The diff is at https://github.com/spack/spack/compare/develop...cosmicexplorer:convert-cc-to-posix-sh-awk-version?expand=1, and the timing from configuring `gettext` was:
```bash
Time =  80.69439458847046
```

It's possible this could be improved further, but it appears to be slightly less performant than the *current* version of this PR after switching from `[ ... ]` to `case` statements (which is at around 78 seconds as per https://github.com/spack/spack/pull/26048#issuecomment-922761447). I was hoping that using (POSIX) `awk`'s native support for arrays would improve the time over (POSIX) `sh` string manipulations slightly, but that doesn't appear to immediately be the case.

However, I haven't yet tried to optimize the `awk` code yet. I'm going to continue with this work by first ignoring the `awk` version, and then:
1. removing some of the parts like processing of `PATH` and environment variables which don't change run-over-run within a `./configure` execution, and do that within python instead,
2. running `cc` with `set -x` and seeing if anything surprising is happening,
3. trying to optimize the `awk` version a little.

tgamblin committed:
Remove UNSET_CC_ENVIRONMENT_VARIABLE stuff

We don't need this -- it's really only there for two arguments, and empty string is just
as good for them.

tgamblin committed:
remove echo override (it's not called)

tgamblin commented:
Closing to focus on #26259.
